programming_examples/matrix_scalar_add/matrix_scalar_add_herd.py

```diff
@@ -0,0 +1,95 @@
+# Copyright (C) 2024, Advanced Micro Devices, Inc.
+# SPDX-License-Identifier: MIT
+
+from air.ir import *
+from air.dialects.air import *
+from air.dialects import arith
+from air.dialects.memref import load, store
+from air.dialects.func import FuncOp
+from air.dialects.scf import for_, yield_
+
+from common import *
+
+range_ = for_
+
+
+@module_builder
+def build_module():
+    memrefTyInOut = MemRefType.get(IMAGE_SIZE, T.i32())
+
+    # Create two channels which will send/receive the
+    # input/output data respectively
+    ChannelOp("ChanIn")
+    ChannelOp("ChanOut")
+
+    # We will send the image worth of data in and out
+    @FuncOp.from_py_func(memrefTyInOut, memrefTyInOut)
+    def addone(arg0, arg1):
+
+        # The arguments are the input and output
+        @launch(operands=[arg0, arg1])
+        def launch_body(a, b):
+
+            # Read/write the data regions represented by the parameters
+            # into/out of the respective channels.
+            ChannelPut("ChanIn", a)
+            ChannelGet("ChanOut", b)
+
+            @segment(name="seg")
+            def segment_body():
+
+                # The herd sizes correspond to the dimensions of the
+                # contiguous block of cores we are hoping to get.
+                @herd(
+                    name="addherd",
+                    sizes=[
+                        # In this case, we map our data tiles to cores directly.
+                        IMAGE_SIZE[0] // TILE_SIZE[0],
+                        IMAGE_SIZE[1] // TILE_SIZE[1],
+                    ],
+                )
+                def herd_body(tx, ty, sx, sy):
+
+                    # We want to store our data in L1 memory
+                    mem_space = IntegerAttr.get(T.i32(), MemorySpace.L1)
+
+                    # This is the type definition of the tile
+                    tile_type = MemRefType.get(
+                        shape=TILE_SIZE,
+                        element_type=T.i32(),
+                        memory_space=mem_space,
+                    )
+
+                    # We must allocate a buffer of the tile size for the input/output
+                    tile_in = Alloc(tile_type)
+                    tile_out = Alloc(tile_type)
+
+                    # Input a tile
+                    ChannelGet("ChanIn", tile_in)
```

Comment:
@jgmelber I'm not sure how much to rely on implicit behavior in this example. What I have here seems to work, but should I instead explicitly specify how to walk through the data using the strides/sizes/offsets in the channel? What is (or should be) considered best practice?